### PR TITLE
DE461660 : Environment dependencies are missing while export

### PR DESCRIPTION
**Pull request by RaghuReddy**
## Description
Fixed the environment dependencies issue

## Links
Link Type   | Link
------      | ------
Rally Issue | DE461660
Func Spec   | 
## Checklist
- [x] Pull Request Created
- [ ] Code Review Completed
- [ ] Veracode scan results addressed
- [ ] SonarQube scan results addressed
- [ ] TPSR has been submitted (if applicable) 
- [X] Unit tests have been added
- [ ] Integration/Functional test cases have been written and automated
- [ ] Func Spec has been updated as needed
- [ ] Rally Issue(s) are in an engineering completed state

### CODE REVIEW

File: gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/writer/PolicyWriter.java [hunk 1]
```diff
@@ -140,23 +141,41 @@ private void writePolicyMetadata(final Map<String, PolicyMetadata> policyMetadat
...
+    private void populateDependencies(Map<Dependency, List<Dependency>> dependencyListMap, String rootId, String id, Set<Dependency> dependencies) {
+        Set<Map.Entry<Dependency, List<Dependency>>> entrySet = dependencyListMap.entrySet();
+        for (Map.Entry<Dependency, List<Dependency>> entry : entrySet) {
+            Dependency parent = entry.getKey();
+            if (parent.getId().equals(id)) {
+                List<Dependency> dependencyList = entry.getValue();
+                for (Dependency dependency : dependencyList) {
+                    if (!rootId.equals(dependency.getId()) && dependencies.add(dependency) && !BuilderConstants.NON_ENV_ENTITY_TYPES.contains(dependency.getType())) {
+                        populateDependencies(dependencyListMap, rootId, dependency.getId(), dependencies);
+                    }
```
Comment (nikhilahuja) [inline gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/writer/PolicyWriter.java:175]: You can make this logic simple to read by moving out the recursive call for Env entities outside of this method. Like,

```
for (Dependency dependency : dependencyList) {
    if (!NON_ENV_ENTITY_TYPES.contains(dependency.getType())) {
        populateEnvTransitiveDependencies();
    } else {
        dependencies.add(dependency);
    }
}

private void populateEnvTransitiveDependencies() { ... }
```

Comment (RaghuReddy) [reply gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/writer/PolicyWriter.java]: @nikhilahuja , why do we need separate method, that looks similar to existing method.
If it is just only for readability we can add proper comment over it.

File: gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/writer/PolicyWriter.java [hunk 2]
```diff
@@ -140,23 +141,41 @@ private void writePolicyMetadata(final Map<String, PolicyMetadata> policyMetadat
...
+     */
+    private void populateDependencies(Map<Dependency, List<Dependency>> dependencyListMap, String id, Set<Dependency> dependencies) {
+        Set<Map.Entry<Dependency, List<Dependency>>> entrySet = dependencyListMap.entrySet();
+        for (Map.Entry<Dependency, List<Dependency>> entry : entrySet) {
+            Dependency parent = entry.getKey();
+            if (parent.getId().equals(id)) {
+                List<Dependency> dependencyList = entry.getValue();
+                for (Dependency dependency : dependencyList) {
+                    // add the dependency and populate the transitive dependencies for only environmental entities
+                    if (dependencies.add(dependency) && !BuilderConstants.NON_ENV_ENTITY_TYPES.contains(dependency.getType())) {
```
Comment (uppoju) [inline gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/writer/PolicyWriter.java:173]: ENV ENTITY TYPES conditions should be first i guess

File: gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/writer/PolicyWriter.java [hunk 3]
```diff
@@ -140,23 +141,41 @@ private void writePolicyMetadata(final Map<String, PolicyMetadata> policyMetadat
...
+                        populateDependencies(dependencyListMap, dependency.getId(), dependencies);
```
Comment (uppoju) [inline gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/writer/PolicyWriter.java:174]: Raju was saying we should not have transitive dependencies, please check with him

File: gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/writer/PolicyWriter.java [hunk 4]
```diff
@@ -140,23 +141,41 @@ private void writePolicyMetadata(final Map<String, PolicyMetadata> policyMetadat
...
-                Dependency parent = entry.getKey();
-                if (parent.getId().equals(id)) { // Search for "id" to get its dependencies
-                    return new HashSet<>(entry.getValue());
+            populateDependencies(dependencyListMap, id, dependencyList);
+        }
+        return dependencyList;
+    }
+
+    /**
+     * This method recursively loads the entity dependencies from the given entity id
```
Comment (nikhilahuja) [inline gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/writer/PolicyWriter.java:159]: This needs modification if you are going to use the same method recursively only for Env entities.

File: gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/writer/PolicyWriter.java (hunk 3 above)
Comment (RaghuReddy) [reply gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/writer/PolicyWriter.java:174]: @uppoju , I discussed with him, we should not have transitive dependencies for non-environmental entities

File: gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/writer/PolicyWriter.java (hunk 2 above)
Comment (RaghuReddy) [reply gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/writer/PolicyWriter.java:173]: @uppoju , that way we will miss the direct dependencies (non-environmental entities)

File: gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/writer/PolicyWriter.java (hunk 3 above)
Comment (uppoju) [reply gateway-export-plugin/src/main/java/com/ca/apim/gateway/cagatewayexport/tasks/explode/writer/PolicyWriter.java:174]: ok

**Comment by RaghuReddy** [comment]
created new PR https://github.com/CAAPIM/gateway-developer-plugin/pull/251 as base branch is no longer valid

**Pull request closed without merging**